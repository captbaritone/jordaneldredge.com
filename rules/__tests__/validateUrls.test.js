const path = require("path");
const fs = require("fs");
const fetch = require("node-fetch");

async function urlExists(url, timeoutms) {
  const controller = new AbortController();
  const timeout = setTimeout(() => {
    controller.abort();
  }, timeoutms); // will time out after 1000ms
  try {
    const response = await fetch(url, {
      signal: controller.signal,
      method: "HEAD",
    });
    if (response.status !== 200) {
      console.log(response.status, url);
    }
    return response.status === 200;
  } catch {
    return false;
  } finally {
    clearTimeout(timeout);
  }
}

function readJson(filePath) {
  const fileContent = fs.readFileSync(filePath, { encoding: "utf8" });
  return JSON.parse(fileContent || "[]");
}

async function withJson(filePath, cb) {
  const json = readJson(filePath);
  const newJson = await cb(json);
  const newFileContent = JSON.stringify(newJson, null, 2);
  fs.writeFileSync(filePath, newFileContent);
}

test("Validate image URLs", async () => {
  const imagesPath = path.join(__dirname, "..", "images.json");
  await withJson(imagesPath, async (existing) => {
    const todo = Object.entries(existing).filter(
      ([key, value]) => value === "unverified"
    );

    const jobs = todo.map(async ([key, value]) => {
      const url = "https://jordaneldredge.com" + key;
      const exists = await urlExists(url, 1000);
      existing[key] = exists ? "exists" : "notfound";
    });

    await Promise.all(jobs);
    return existing;
  });

  const existing = readJson(imagesPath);
  const values = Object.values(existing);
  const all = values.every((v) => v === "exists");
  expect(all).toBe(true);
});

test("Validate link URLs", async () => {
  const linksPath = path.join(__dirname, "..", "links.json");
  await withJson(linksPath, async (existing) => {
    const todo = Object.entries(existing)
      .filter(([key, value]) => value === "unverified")
      .slice(0, 50);

    const jobs = todo.map(async ([key, value]) => {
      const exists = await urlExists(key, 1000);
      existing[key] = exists ? "exists" : "notfound";
    });

    await Promise.all(jobs);

    return existing;
  });

  const existing = readJson(linksPath);
  const invalidUrls = Object.entries(existing)
    .filter(([key, value]) => value !== "exists")
    .map(([key]) => key);
  invalidUrls.sort();
  expect(invalidUrls).toMatchSnapshot();
});
