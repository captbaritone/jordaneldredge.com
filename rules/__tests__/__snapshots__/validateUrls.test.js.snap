// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate link URLs 1`] = `
Array [
  "http://banach.ucsd.edu/RefreshCannon/",
  "http://banach.ucsd.edu/RefreshCannon/wins.php",
  "http://beach-boys.aure.com/",
  "http://bit.ly",
  "http://blog.classicalcode.com/projects/shoot/high_score.php",
  "http://blog.classicalcode.com/projects/shoot/register.php",
  "http://classicalcode.com/chelseahollow/?cat=9",
  "http://classicalcode.com/projects/hero/high_score.php",
  "http://classicalcode.com/projects/hero/register.php",
  "http://codeception.com/docs/03-ModulesAndHelpers",
  "http://en.wikipedia.org/wiki/Combination_tone",
  "http://en.wikipedia.org/wiki/Der_Erlk%C3%B6nig",
  "http://en.wikipedia.org/wiki/Forever_Plaid",
  "http://en.wikipedia.org/wiki/Golden_Gate_Park",
  "http://en.wikipedia.org/wiki/Harmonic_series_(music)",
  "http://en.wikipedia.org/wiki/Overtone",
  "http://en.wikipedia.org/wiki/Robert_Schumann",
  "http://en.wikipedia.org/wiki/Schadenfreude",
  "http://en.wikipedia.org/wiki/Tag_%28Barbershop%29",
  "http://en.wikipedia.org/wiki/Tom_Lehrer",
  "http://en.wikipedia.org/wiki/War_(card_game)",
  "http://facepwn.com/avatarstats.php",
  "http://fileformats.archiveteam.org/wiki/RIFF",
  "http://forums.macrumors.com/showthread.php?t=223370",
  "http://gnupg.org/",
  "http://illinistatesmen.org/files/classic_tags.pdf",
  "http://maps.google.com/maps?f=q&hl=en&geocode=&q=sf&ie=UTF8&ll=37.784605,-122.458119&spn=0.006817,0.013711&z=16&layer=c&cbll=37.7812,-122.45887&cbp=1,566.3613579897253,,0,7.719995925626088",
  "http://music.classicalcode.com",
  "http://musicdance.sfsu.edu/faculty/84/david-xiques",
  "http://sacrej.eu/index.php/2009/05/06/refresh-cannon/",
  "http://thatsmathematics.com/",
  "http://tinyurl.com/nl92nr8",
  "http://tinyurl.com/porqfo9",
  "http://translate.google.com/translate?u=http%3A%2F%2Fwww.superlevel.de%2F&sl=de&tl=en&hl=en&ie=UTF-8",
  "http://us2.php.net/manual/en/function.session-start.php",
  "http://us2.php.net/manual/en/function.session-write-close.php",
  "http://vimeo.com",
  "http://vimeo.com/2042526",
  "http://vimeo.com/user863964",
  "http://webassembly.Global",
  "http://winterreise.classicalcode.com/",
  "http://winterreise.classicalcode.com/?feed=rss2",
  "http://www.16thmission.com/",
  "http://www.acdaonline.org/",
  "http://www.amazon.com/Zoom-ZOO-H4-Handy-Recorder/dp/B000LGA2K6",
  "http://www.amazon.com/gp/product/B000LGA2K6?ie=UTF8&tag=winterjourna-20&linkCode=as2&camp=1789&creative=390957&creativeASIN=B000LGA2K6",
  "http://www.classicalcode.com/t.php",
  "http://www.classicalcode.com/wiki/",
  "http://www.classicsforkids.com/teachers/training/handsigns.asp",
  "http://www.davescooltoys.com/twists.html",
  "http://www.fmylife.com",
  "http://www.foundmagazine.com",
  "http://www.funnytimes.com/",
  "http://www.godaddy.com",
  "http://www.hpmendoza.com/",
  "http://www.limerickdb.com",
  "http://www.math.ucsd.edu/~neldredg/",
  "http://www.meetup.com/jsmeetup/",
  "http://www.meetup.com/vim-sf/",
  "http://www.operaworks.com/",
  "http://www.pcmag.com/article2/0,1759,2243715,00.asp",
  "http://www.pgp.com/",
  "http://www.sadtrombone.com",
  "http://www.sfsu.edu/~music/patty/faculty.html#deeter",
  "http://www.socratesrules.com",
  "http://www.spypig.com/",
  "http://www.stampedecitychorus.com/classic_tags_men2.pdf",
  "http://www.theflip.com/products_flip_ultra.shtml",
  "http://www.thesublimelight.com",
  "http://www.woot.com/",
  "http://www.youtube.com/user/VideoBuck",
  "http://www.youtube.com/watch?v=4mk-AtNLfrM",
  "http://www.youtube.com/watch?v=BtsjKPqKgzg",
  "http://www.youtube.com/watch?v=DUkWlxyV4Zw",
  "http://www.youtube.com/watch?v=L8CwvvDzLC4",
  "http://www.youtube.com/watch?v=P5B6nysheec",
  "http://www.youtube.com/watch?v=Q1ljOdTM9l4",
  "http://www.youtube.com/watch?v=YKK5I9_LFZM&feature=related",
  "http://www.youtube.com/watch?v=imHoaelEzOc&feature=related",
  "http://www.youtube.com/watch?v=kFWYSW4vfcA&feature=related",
  "http://www.zaynardstudios.com/Welcome.html",
  "http://www.zoneedit.com",
  "https://aiap.it/prodotto/progetto-grafico-37-2/",
  "https://aiap.it/prodotto/progetto-grafico-37/",
  "https://archive.org/",
  "https://archive.org/details/TOPCUR95_ZIP",
  "https://blog.archive.org/2018/10/02/dont-click-on-the-llama/",
  "https://bugs.webkit.org/show_bug.cgi?id=221589",
  "https://codesandbox.io/s/ani-web-animation-lz46u?file=/src/parseAni.js",
  "https://creativecommons.org/licenses/by-sa/4.0/",
  "https://css-tricks.com/forums/topic/animated-cursor/",
  "https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Basic_User_Interface/Using_URL_values_for_the_cursor_property",
  "https://discord.com/",
  "https://drafts.csswg.org/web-animations-1/#discrete",
  "https://filext.com/file-extension/ANI",
  "https://gist.github.com/4093068",
  "https://github.com/captbaritone",
  "https://github.com/captbaritone/eldredge-lazy_afternoon",
  "https://github.com/captbaritone/eldredge-shepard_tone",
  "https://github.com/captbaritone/eldredge-thats_why_i_love_you",
  "https://github.com/captbaritone/eslint-plugin-underscore/blob/master/docs/rules/prefer-compact.md",
  "https://github.com/captbaritone/eslint-plugin-underscore/blob/master/docs/rules/prefer-reject.md",
  "https://github.com/captbaritone/hashb.in",
  "https://github.com/captbaritone/webamp/tree/master/packages/ani-cursor",
  "https://github.com/captbaritone/winamp-skins/raw/master/v2/base-2.91.wsz",
  "https://github.com/captbaritone/winamp2-js/blob/master/js/skinParser.js",
  "https://github.com/epistemex/cardinal-spline-js",
  "https://github.com/imagemin/imagemin-optipng",
  "https://github.com/rochars/byte-data",
  "https://github.com/rochars/riff-file",
  "https://hearsaysocial.com/company/careers/engineering/",
  "https://jordaneldredge.com/blog/winamp-skin-musuem",
  "https://marketplace.visualstudio.com/items?itemName=JordanEldredge.implicit-parentheses",
  "https://motherboard.vice.com/en_us/article/qvebbv/winamp-2-mp3-music-player-emulator",
  "https://news.ycombinator.com/item?id=15314629",
  "https://news.ycombinator.com/item?id=16333550",
  "https://news.ycombinator.com/item?id=17583997",
  "https://news.ycombinator.com/item?id=24373699",
  "https://news.ycombinator.com/item?id=8565665",
  "https://nsfwjs.com/",
  "https://pragprog.com/book/tpp/the-pragmatic-programmer",
  "https://raw.githubusercontent.com/captbaritone/programming-blog-content/master/blog/bingy-bongy-music-in-javascript.md",
  "https://raw.githubusercontent.com/captbaritone/programming-blog-content/master/blog/creating-the-shepard-tone-audio-illusion-with-javascript.md",
  "https://raw.githubusercontent.com/captbaritone/programming-blog-content/master/blog/drawing-mandalas-with-javascript-and-canvas.md",
  "https://skins.webamp.org/skin/8820f5a444d7eeedb6e610f117900eac/TealAmp_winamp.zip/",
  "https://techcrunch.com/2018/02/09/whip-the-llamas-ass-with-this-javascript-winamp-emulator/",
  "https://towardsdatascience.com/run-stylegan2-ada-on-an-aws-spot-instance-in-no-time-d2022fc1e119",
  "https://trac.webkit.org/changeset/269812/webkit",
  "https://twitter.com/aadsm",
  "https://twitter.com/luigihann?lang=en",
  "https://twitter.com/winampskins/",
  "https://webamp.org/?skinUrl=https://cdn.webampskins.org/skins/4308a2fc648033bf5fe7c4d56a5c8823.wsz",
  "https://webamp.org/?skinUrl=https://cdn.webampskins.org/skins/4cbfadd11c0e8ebec834ea0355d275c1.wsz",
  "https://webamp.org/?skinUrl=https://cdn.webampskins.org/skins/6e30f9e9b8f5719469809785ae5e4a1f.wsz",
  "https://www.1014.org/",
  "https://www.algolia.com/",
  "https://www.cloudflare.com/",
  "https://www.facebook.com/rsrc.php/v2/yc/r/J_2KFGBpaBp.png",
  "https://www.linkedin.com/in/jordaneldredge/",
  "https://www.netlify.com/",
  "https://www.netlify.com/pricing/",
]
`;
